subsys/bluetooth/host/att.c

```diff
@@ -2948,6 +2985,53 @@ static int bt_att_accept(struct bt_conn *conn, struct bt_l2cap_chan **ch)
 BT_L2CAP_CHANNEL_DEFINE(att_fixed_chan, BT_L2CAP_CID_ATT, bt_att_accept, NULL);
 
 #if defined(CONFIG_BT_EATT)
+static k_timeout_t credit_based_connection_delay(struct bt_conn *conn)
+{
+	/*
+	 * 5.3 Vol 3, Part G, Section 5.4 L2CAP COLLISION MITIGATION
+	 * ... In this situation, the Central may retry
+	 * immediately but the Peripheral shall wait a minimum of 100 ms before retrying;
+	 * on LE connections, the Peripheral shall wait at least 2 *
+	 * (connPeripheralLatency + 1) * connInterval if that is longer.
+	 */
+
+	if (conn->role == BT_CONN_ROLE_CENTRAL) {
+		return K_NO_WAIT;
+	}
+
+	const uint32_t rand_delay = sys_rand32_get() & 0x7; /* Small random delay for IOP */
+	const uint32_t calculated_delay =
+		2 * (conn->le.latency + 1) * BT_INTERVAL_TO_MS(conn->le.interval);
+
+	return K_MSEC(MAX(100, calculated_delay + rand_delay));
+}
+
+int att_schedule_eatt_connect(struct bt_conn *conn, uint8_t eatt_chans_to_connect)
+{
+	struct bt_att *att = att_get(conn);
+
+	att->eatt_chans_to_connect = eatt_chans_to_connect;
+	return k_work_schedule(&att->connection_work, credit_based_connection_delay(conn));
```

Comment:
It is a bit risky to return `k_work_schedule` as that doesn't return 0 on success and negative otherwise (which is of course typical):

```c
 * @retval 0 if work was already scheduled or submitted.
 * @retval 1 if work has been scheduled.
 * @retval -EBUSY if @p delay is @c K_NO_WAIT and
 *         k_work_submit_to_queue() fails with this code.
 * @retval -EINVAL if @p delay is @c K_NO_WAIT and
 *         k_work_submit_to_queue() fails with this code.
 * @retval -ENODEV if @p delay is @c K_NO_WAIT and
 *         k_work_submit_to_queue() fails with this code.
```

So both 0 and 1 can be returned as success. 

Furthermore, should we call `k_work_reschedule` instead, or do we never risk the work already being scheduled?

Reply:
`k_work_reschedule()` fits better with the requirement in the spec of waiting at least a given time. Though I do not think it is possible for this to be already scheduled.

Currently the return value from this function is ignored in `ecred_connect_cb()` as I am not sure if there is anything one can do in the case of failure. We are still spec compliant as no connection requests are sent.

Reply:
I am personally opposed to have things fail silently, so at the very least the error code should be logged, but best to somehow forward it to the original caller.

Reply:
Yes, failing silently is not good, will log an error here. I don't see what the L2CAP layer can do if the scheduling of the retry of the connection fails.

Reply:
That's a good question. I am not sure what L2CAP can do either, besides trying again, but that seems unlikely to work.